keymanager/src/client/remote.rs

```diff
@@ -244,33 +245,46 @@ impl RemoteClient {
         let mut shares = Vec::with_capacity(min_shares);
 
         // Fetch key shares in random order.
-        // TODO: Optimize by fetching key shares concurrently.
         let mut committee = status.committee;
         committee.shuffle(&mut OsRng);
 
-        for node_id in committee {
-            // Stop fetching when enough shares are received.
-            if shares.len() == min_shares {
-                break;
+        // Fetch key shares concurrently.
+        let mut futures = FuturesUnordered::new();
+
+        loop {
+            // Continuously add new key share requests until the required
+            // number of key shares is received, ensuring the future queue
+            // remains filled even if some requests fail.
+            while shares.len() + futures.len() < min_shares {
+                let node_id = match committee.pop() {
+                    Some(node_id) => node_id,
+                    None => return Err(KeyManagerError::InsufficientKeyShares),
+                };
+
+                let future = self
+                    .rpc_client
+                    .secure_call(
+                        METHOD_SGX_POLICY_KEY_SHARE,
+                        KeyShareRequest {
+                            id: status.id,
+                            runtime_id: status.runtime_id,
+                            epoch: status.handoff,
+                            key_runtime_id: self.runtime_id,
+                            key_id,
+                        },
+                        vec![node_id],
+                    )
+                    .await
```

Comment:
This seems like it is awaiting for the call the complete before proceeding? The `into_result_with_feedback` should be done further below after the call future resolves.